lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+// hw.array-typed 'array'.
+static Value narrowValueToArrayWidth(OpBuilder &builder, Value array,
+                                     Value value) {
+  OpBuilder::InsertionGuard g(builder);
+  if (value.isa<BlockArgument>())
+    builder.setInsertionPointToStart(value.getParentBlock());
+  else
+    builder.setInsertionPointAfter(value.getDefiningOp());
+  auto arrayType = array.getType().cast<hw::ArrayType>();
+  return builder.create<comb::ExtractOp>(
+      value.getLoc(), value, /*lowBit=*/0,
+      llvm::Log2_64_Ceil(arrayType.getSize()));
+}
+
+struct EliminateParamValueOpPattern : public OpRewritePattern<ParamValueOp> {
+  EliminateParamValueOpPattern(MLIRContext *context, ArrayAttr parameters)
+      : OpRewritePattern<ParamValueOp>(context), parameters(parameters) {}
+
+  LogicalResult matchAndRewrite(ParamValueOp op,
+                                PatternRewriter &rewriter) const override {
+    // Substitute the param value op with an evaluated constant operation.
+    auto paramValue =
+        evaluateParametricAttr(op.getLoc(), parameters, op.value());
+    if (failed(paramValue))
+      return failure();
+    rewriter.replaceOpWithNewOp<hw::ConstantOp>(
+        op, op.getType(), paramValue.getValue().getSExtValue());
+    return success();
+  }
+
+  ArrayAttr parameters;
+};
+
+// hw.array_get operations require indexes to be of equal width of the
+// array itself. Since indexes may originate from constants or parameters, emit
+// comb.extract operations to fulfil this invariant.
+struct NarrowArrayGetIndexPattern : public OpConversionPattern<ArrayGetOp> {
+public:
+  using OpConversionPattern<ArrayGetOp>::OpConversionPattern;
+
+  LogicalResult
+  matchAndRewrite(ArrayGetOp op, OpAdaptor adaptor,
+                  ConversionPatternRewriter &rewriter) const override {
+    auto inputType = type_cast<ArrayType>(op.input().getType());
+    Type targetIndexType = IntegerType::get(
+        getContext(),
+        inputType.getSize() == 1 ? 1 : llvm::Log2_64_Ceil(inputType.getSize()));
+
+    if (op.index().getType().getIntOrFloatBitWidth() ==
+        targetIndexType.getIntOrFloatBitWidth())
+      return failure(); // nothing to do
+
+    // Narrow the index value.
+    auto narrowedIndex =
+        narrowValueToArrayWidth(rewriter, op.input(), op.index());
+    rewriter.replaceOpWithNewOp<ArrayGetOp>(op, op.input(), narrowedIndex);
+    return success();
+  }
+};
+
+// Generic pattern to convert parametric result types.
+struct ParametricTypeConversionPattern : public ConversionPattern {
+  ParametricTypeConversionPattern(MLIRContext *ctx,
+                                  TypeConverter &typeConverter,
+                                  ArrayAttr parameters)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), /*benefit=*/1,
+                          ctx),
+        parameters(parameters) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
```

Comment:
nit: I think there is a new thing call `OpAdapter` and the `ArrayRef<Value>` form is deprecated. If so, please use the new form for new conversion patterns.

Reply:
`OpAdaptor` is only available for `OpConversionPattern`s - this is a `ConversionPattern` which only provides the old form.

Reply:
I see, sounds good, sorry for the confusion.